"use strict";(self.webpackChunkada_lang_io=self.webpackChunkada_lang_io||[]).push([[7344],{6549:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>A,contentTitle:()=>h,default:()=>x,frontMatter:()=>m,metadata:()=>f,toc:()=>k});var o=n(1716),a=n(3183),r=Object.defineProperty,i=Object.defineProperties,s=Object.getOwnPropertyDescriptors,c=Object.getOwnPropertySymbols,l=Object.prototype.hasOwnProperty,p=Object.prototype.propertyIsEnumerable,d=(e,t,n)=>t in e?r(e,t,{enumerable:!0,configurable:!0,writable:!0,value:n}):e[t]=n,u=(e,t)=>{for(var n in t||(t={}))l.call(t,n)&&d(e,n,t[n]);if(c)for(var n of c(t))p.call(t,n)&&d(e,n,t[n]);return e};const m={sidebar_position:86},h="11 Exceptions",f={unversionedId:"arm/AA-11/AA-11",id:"arm/AA-11/AA-11",title:"11 Exceptions",description:"We're still working on the Reference manual output.  Internal links are broken,",source:"@site/docs/arm/AA-11/AA-11.mdx",sourceDirName:"arm/AA-11",slug:"/arm/AA-11/",permalink:"/docs/arm/AA-11/",draft:!1,tags:[],version:"current",sidebarPosition:86,frontMatter:{sidebar_position:86},sidebar:"tutorialSidebar",previous:{title:"10.2 Program Execution",permalink:"/docs/arm/AA-10/AA-10.2"},next:{title:"11.1 Exception Declarations",permalink:"/docs/arm/AA-11/AA-11.1"}},A={},k=[{value:"Wording Changes from Ada 83",id:"wording-changes-from-ada-83",level:4},{value:"Wording Changes from Ada 2005",id:"wording-changes-from-ada-2005",level:4}],g={toc:k};function x(e){var t,n=e,{components:r}=n,d=((e,t)=>{var n={};for(var o in e)l.call(e,o)&&t.indexOf(o)<0&&(n[o]=e[o]);if(null!=e&&c)for(var o of c(e))t.indexOf(o)<0&&p.call(e,o)&&(n[o]=e[o]);return n})(n,["components"]);return(0,o.kt)("wrapper",(t=u(u({},g),d),i(t,s({components:r,mdxType:"MDXLayout"}))),(0,o.kt)("h1",u({},{id:"11-exceptions"}),"11 Exceptions"),(0,o.kt)("admonition",u({},{type:"warning"}),(0,o.kt)("p",{parentName:"admonition"},"We're still working on the Reference manual output.  Internal links are broken,\nas are a bunch of other things.\nSee the ",(0,o.kt)("a",u({parentName:"p"},{href:"https://github.com/ada-lang-io/ada-lang-io/issues/20"}),"tracking issue"))),(0,o.kt)("p",null,"{",(0,o.kt)("em",null,"AI05-0299-1"),"}"," [This clause defines the facilities for dealing with errors or other exceptional situations that arise during program execution.] An ",(0,o.kt)("em",null,"exception")," represents a kind of exceptional situation; an occurrence of such a situation (at run time) is called an ",(0,o.kt)("em",null,"exception occurrence"),". [ To ",(0,o.kt)("em",null,"raise")," an exception is to abandon normal program execution so as to draw attention to the fact that the corresponding situation has arisen. Performing some actions in response to the arising of an exception is called ",(0,o.kt)("em",null,"handling")," the exception. ]"),(0,o.kt)(a.Z,{type:"note",mdxType:"Admonition"},(0,o.kt)("p",null,(0,o.kt)("strong",null,"To be honest: "),"...or handling the exception occurrence. ")),(0,o.kt)(a.Z,{type:"note",mdxType:"Admonition"},(0,o.kt)("p",null,(0,o.kt)("strong",null,"Ramification: "),"For example, an exception End_Error might represent error situations in which an attempt is made to read beyond end-of-file. During the execution of a partition, there might be numerous occurrences of this exception. ")),(0,o.kt)(a.Z,{type:"note",mdxType:"Admonition"},(0,o.kt)("p",null,(0,o.kt)("strong",null,"To be honest: "),"{",(0,o.kt)("em",null,"AI12-0426-1"),"}",' When the meaning is clear from the context, we sometimes use "',(0,o.kt)("em",null,"occurrence"),'" as a shorthand for "exception occurrence". ')),(0,o.kt)("p",null,"Version=[5],Kind=(AddedNormal),Group=[E],Term=[exception], Def=[a kind of exceptional situation] Version=[5],Kind=(AddedNormal),Group=[E],Term=[exception occurrence], Def=[a run-time occurrence of an exceptional situation]"),(0,o.kt)("p",null,"{",(0,o.kt)("em",null,"AI05-0043-1"),"}"," ","{",(0,o.kt)("em",null,"AI05-0258-1"),"}"," [An ",(0,o.kt)("code",null,(0,o.kt)("a",{href:"../AA-11/AA-11.1#S0303"},"exception_declaration"))," declares a name for an exception. An exception can be raised explicitly (for example, by a ",(0,o.kt)("code",null,(0,o.kt)("a",{href:"../AA-11/AA-11.3#S0308"},"raise_statement")),") or implicitly (for example, by the failure of a language-defined check). When an exception arises, control can be transferred to a user-provided ",(0,o.kt)("code",null,(0,o.kt)("a",{href:"../AA-11/AA-11.2#S0305"},"exception_handler"))," at the end of a ",(0,o.kt)("code",null,(0,o.kt)("a",{href:"../AA-11/AA-11.2#S0304"},"handled_sequence_of_statements")),", or it can be propagated to a dynamically enclosing execution.] "),(0,o.kt)("h4",u({},{id:"wording-changes-from-ada-83"}),"Wording Changes from Ada 83"),(0,o.kt)(a.Z,{type:"note",mdxType:"Admonition"},(0,o.kt)("p",null,"We are more explicit about the difference between an exception and an occurrence of an exception. This is necessary because we now have a type (Exception_Occurrence) that represents exception occurrences, so the program can manipulate them. Furthermore, we say that when an exception is propagated, it is the same occurrence that is being propagated (as opposed to a new occurrence of the same exception). The same issue applies to a re-raise statement. In order to understand these semantics, we have to make this distinction. ")),(0,o.kt)("h4",u({},{id:"wording-changes-from-ada-2005"}),"Wording Changes from Ada 2005"),(0,o.kt)(a.Z,{type:"note",mdxType:"Admonition"},(0,o.kt)("p",null,"{",(0,o.kt)("em",null,"AI05-0043-1"),"}"," ",(0,o.kt)("strong",null,"Correction:")," Generalized the introductory description of how an exception can be raised so that it does not appear to cover all of the cases. ")))}x.isMDXComponent=!0}}]);